Use `Lightning CSS` in the `PostCSS` Plugin

RobinMalfait opened the pull request:
This PR will add `Lightning CSS` to the `PostCSS` Plugin in the `oxide` engine. This also required a
little bit of fiddling around with tests because the `stable` version doesn't use that.

To solve this, we will run both version through `Lightning CSS` but purely in tests to get the same
output. It technically makes things a bit slower (but only in tests) and they are now also
consistent.

There is a confusing part though, if you use something like this:
```js
expect(output).toMatchFormattedCss(css`
  .example {
    color: rgb(0, 0, 0);
  }
`)
```

Then the test will pass, but logging the `output` could look like:
```css
.example {
  color: black;
}
```

Not sure if this is an _actual_ problem right now or not...

<!--

👋 Hey, thanks for your interest in contributing to Tailwind!

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a lot of time and effort into a new feature. To avoid this from happening, we request that contributors create an issue to first discuss any significant new features. This includes things like adding new utilities, creating new at-rules, or adding new component examples to the documentation.

https://github.com/tailwindcss/tailwindcss/blob/master/.github/CONTRIBUTING.md

-->

RobinMalfait committed:
use `lightningcss` in the main PostCss Plugin

RobinMalfait committed:
use lightningcss in our custom matchers

Now that we are using `lightningcss` and nesting in the new `oxide`
engine, the generated output _will_ be different in the majority of test
cases.

Using a combination of `prettier` and `lightningcss` will make the
output consistent.

The moment we are fully using the `oxide` engine, we can drop
`lightningcss` or `prettier` again to improve the performance of the
tests.

RobinMalfait committed:
update tests to apply `lightningcss` related changes

RobinMalfait committed:
bump lightningcss

RobinMalfait committed:
update changelog

thecrypticace commented on tests/arbitrary-variants.test.js:
```diff
@@ -643,18 +643,18 @@ crosscheck(({ stable, oxide }) => {
+              <div class="group-aria-[labelledby="a_b"]:underline"></div>
+              <div class="peer-aria-[labelledby="a_b"]:underline"></div>
+              <div class="group-aria-[sort=:"ascending"]/foo:underline"></div>
+              <div class="peer-aria-[sort="ascending"]/foo:underline"></div>
```
While this technically works in the test it would be invalid HTML. Should you change to single quotes?

RobinMalfait replied:
Yep thanks! I removed the quotes again in most cases, but added them for the once where it was required like `foo_bar` cases.

thecrypticace commented on jest/customMatchers.js:
```diff
@@ -1,85 +1,132 @@
+      try {
+        // Lightning CSS is pretty strict, so it will fail for `@media screen(md) {}` for example,
+        // in that case we can fallback to prettier since it doesn't really care. However if an
+        // actual sytnax is made, then we still want to show the proper error.
```
actual **syntax error** is made

RobinMalfait committed:
add `lightningcss` and `browserslist` as dev dependencies to stable package.json

RobinMalfait committed:
only use `lightningcss` in tests (without prettier)

We will only fallback to prettier if lightningcss fails somehow.

RobinMalfait committed:
make CI happy (integration tests)

Apply changes to integration tests now that we are using lightningcss

RobinMalfait committed:
transform `lightningcss` for Node 12 when running tests

RobinMalfait committed:
run prettier on failing tests for `toMatchFormattedCss`

This will result in better diffs because diffs are typically per block
and/or per line. But lightningcss will simplify certain selectors and
the diff won't be as clear.

We will only apply the prettier formatting for failing tests in the diff
view so that diffs are cleaner and we don't pay for the additional
prettier calls when tests pass.

RobinMalfait committed:
apply side effect chagnes due to only using lightningcss for tests
